sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala

```diff
@@ -63,6 +65,21 @@ trait Generator extends Expression {
 }
 
 /**
+ * A collection producing [[Generator]]. This trait provides a different path for code generation,
+ * by allowing code generation to return either an [[ArrayData]] or a [[MapData]] object.
+ */
+trait CollectionGenerator extends Generator {
+  /** The position of an element within the collection should also be returned. */
+  def position: Boolean
+
+  /** Rows will be inlined during generation. */
+  def inline: Boolean
+
+  /** The schema of the returned collection object. */
+  def collectionSchema: DataType = dataType
+}
```

Comment:
btw, must we need this interface? Adding a new interface makes codes more complicated, I think.